<center><h1>Multi-Class and Multi-Label Anuran Calls Data Set Classification/Clustering </h1></center>

Name: Lingduo Luo

Github Username: Lingduo-L

Mail: lingduoluo109@gmail.com

In [344]:
import pandas as pd
import numpy as np

import numpy as np
from sklearn.impute import SimpleImputer

from numpy import sqrt 
from tqdm import tqdm
import matplotlib.pyplot as plt

from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics 

from sklearn.metrics import classification_report, confusion_matrix

import os
import seaborn as sns

from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve, auc

import numpy as np
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

from sklearn.svm import SVC
# import numpy as np
from sklearn.multiclass import OneVsRestClassifier
# from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.metrics import hamming_loss
from sklearn.metrics import accuracy_score

from sklearn import preprocessing

from sklearn.svm import LinearSVC
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_classification

from imblearn.over_sampling import SMOTE

import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
%matplotlib inline

from sklearn.metrics import hamming_loss
from scipy.spatial.distance import hamming

### 1. Multi-class and Multi-Label Classification Using Support Vector Machines
### Download the Anuran Calls (MFCCs) Data Set from: https://archive.ics.uci.edu/ml/datasets/Anuran+Calls+%28MFCCs%29. Choose 70% of the data randomly as the training set.

In [2]:
RawData = pd.read_csv('../Data/Anuran Calls (MFCCs)/Frogs_MFCCs.csv')
RawData

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
...        ...       ...       ...       ...       ...       ...       ...   
7190       1.0 -0.554504 -0.337717  0.035533  0.034511  0.443451  0.093889   
7191       1.0 -0.517273 -0.370574  0.030673  0.068097  0.402890  0.096628   
7192       1.0 -0.582557 -0.343237  0.029468  0.064179  0.385596  0.114905   
7193       1.0 -0.519497 -0.307553 -0.004922  0.072865  0.377131  0.086866   
7194       1.0 -0.508833 -0.324106  0.062068  0.078211  0.397188  0.094596   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_17  MFCCs_18  MFCCs_19  \
0    -0.150063 -0.171128  0.124676  ... -0.108351 -0.077623 -0.009568   
1    -0.222475 -0.207693  0.170883  ... -0.090974 -0.056510 -0.035303   
2    -0.242234 -0.219153  0.232538  ... -0.050691 -0.023590 -0.066722   
3    -0.194347 -0.098181  0.270375  ... -0.136009 -0.177037 -0.130498   
4    -0.265423 -0.172700  0.266434  ... -0.048885 -0.053074 -0.088550   
...        ...       ...       ...  ...       ...       ...       ...   
7190 -0.100753  0.037087  0.081075  ...  0.069430  0.071001  0.021591   
7191 -0.116460  0.063727  0.089034  ...  0.061127  0.068978  0.017745   
7192 -0.103317  0.070370  0.081317  ...  0.082474  0.077771 -0.009688   
7193 -0.115799  0.056979  0.089316  ...  0.051796  0.069073  0.017963   
7194 -0.117672  0.058874  0.076180  ...  0.061455  0.072983 -0.003980   

      MFCCs_20  MFCCs_21  MFCCs_22           Family      Genus  \
0     0.057684  0.118680  0.014038  Leptodactylidae  Adenomera   
1     0.020140  0.082263  0.029056  Leptodactylidae  Adenomera   
2    -0.025083  0.099108  0.077162  Leptodactylidae  Adenomera   
3    -0.054766 -0.018691  0.023954  Leptodactylidae  Adenomera   
4    -0.031346  0.108610  0.079244  Leptodactylidae  Adenomera   
...        ...       ...       ...              ...        ...   
7190  0.052449 -0.021860 -0.079860          Hylidae     Scinax   
7191  0.046461 -0.015418 -0.101892          Hylidae     Scinax   
7192  0.027834 -0.000531 -0.080425          Hylidae     Scinax   
7193  0.041803 -0.027911 -0.096895          Hylidae     Scinax   
7194  0.031560 -0.029355 -0.087910          Hylidae     Scinax   

             Species  RecordID  
0     AdenomeraAndre         1  
1     AdenomeraAndre         1  
2     AdenomeraAndre         1  
3     AdenomeraAndre         1  
4     AdenomeraAndre         1  
...              ...       ...  
7190     ScinaxRuber        60  
7191     ScinaxRuber        60  
7192     ScinaxRuber        60  
7193     ScinaxRuber        60  
7194     ScinaxRuber        60  

[7195 rows x 26 columns]

In [3]:
df = RawData.sample(frac=1.0)  # 全部打乱
cut_idx = int(round(0.1 * df.shape[0]))
TestData, TrainData = df.iloc[:int(7195*0.3)], df.iloc[int(7195*0.3):]

In [4]:
print (RawData.shape, TrainData.shape, TestData.shape)

(7195, 26) (5037, 26) (2158, 26)


In [5]:
TrainData

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
4025       1.0  0.409614  0.189652  0.488512  0.153108  0.044635 -0.160591   
977        1.0  0.051639  0.288962  0.368283 -0.042896  0.118380  0.241227   
1117       1.0  0.445827  0.293615  0.405580  0.078879  0.063783  0.125193   
518        1.0  0.549544  0.628990  0.469123 -0.178067 -0.084767  0.460348   
1680       1.0  0.235578  0.150980  0.544132  0.172222 -0.000511 -0.201507   
...        ...       ...       ...       ...       ...       ...       ...   
4981       1.0  0.242088  0.053241  0.404783  0.267120  0.186832  0.063521   
6581       1.0  0.124373  0.324298  0.312395  0.164900  0.240560  0.116508   
5727       1.0  0.163649  0.321016  0.288066  0.152522  0.205293  0.106355   
4878       1.0  0.646854  0.247316  0.461047  0.387061  0.073741 -0.187130   
5064       1.0  0.521036  0.846076  0.485924 -0.179671  0.045003  0.045139   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_17  MFCCs_18  MFCCs_19  \
4025  0.047693  0.336279  0.074517  ...  0.177379  0.000377 -0.119107   
977  -0.034584  0.033428  0.032600  ... -0.087658 -0.044175 -0.023474   
1117  0.023249 -0.153560 -0.033756  ... -0.105378  0.043563  0.016773   
518  -0.006620 -0.345066  0.214803  ...  0.082623  0.166370 -0.096085   
1680 -0.002219  0.328669  0.114075  ...  0.256425  0.131211 -0.095855   
...        ...       ...       ...  ...       ...       ...       ...   
4981 -0.079642 -0.057390  0.271443  ... -0.268151 -0.111373  0.041125   
6581 -0.180234 -0.055447  0.129544  ...  0.151139  0.044809 -0.007393   
5727 -0.120333 -0.030510  0.078287  ...  0.079941  0.006365  0.007588   
4878  0.059394  0.250673  0.101224  ... -0.023786  0.177838  0.087649   
5064  0.086520  0.073915 -0.204901  ... -0.051994 -0.032546 -0.091318   

      MFCCs_20  MFCCs_21  MFCCs_22           Family          Genus  \
4025 -0.150723  0.041190  0.203909  Leptodactylidae      Adenomera   
977  -0.048363  0.040817  0.053208    Dendrobatidae       Ameerega   
1117 -0.072667 -0.010458  0.031258    Dendrobatidae       Ameerega   
518  -0.015425  0.142646  0.039099  Leptodactylidae      Adenomera   
1680 -0.199151  0.012509  0.221572  Leptodactylidae      Adenomera   
...        ...       ...       ...              ...            ...   
4981  0.199325  0.142638 -0.113458          Hylidae  Dendropsophus   
6581 -0.043019 -0.047254  0.036592          Hylidae      Hypsiboas   
5727 -0.001941 -0.048155  0.009939          Hylidae      Hypsiboas   
4878 -0.129743 -0.108685  0.084573          Hylidae  Dendropsophus   
5064  0.135836  0.084960 -0.236737          Hylidae      Hypsiboas   

                     Species  RecordID  
4025  AdenomeraHylaedactylus        23  
977       Ameeregatrivittata        11  
1117      Ameeregatrivittata        13  
518           AdenomeraAndre         8  
1680  AdenomeraHylaedactylus        15  
...                      ...       ...  
4981              HylaMinuta        35  
6581       HypsiboasCordobae        43  
5727       HypsiboasCordobae        40  
4878              HylaMinuta        31  
5064    HypsiboasCinerascens        36  

[5037 rows x 26 columns]

In [6]:
y_train = TrainData.iloc[:,22:25]
X_train = TrainData.drop(['Family','Genus','Species','RecordID'],axis = 1)
# y_train
# X_train

y_test = TestData.iloc[:,22:25]
X_test = TestData.drop(['Family','Genus','Species','RecordID'],axis = 1)

# y_test
# X_test

In [7]:
Family_test = y_test.iloc[:,0:1]
# Family_test
Genus_test = y_test.iloc[:,1:2]
# y_Genus
Species_test = y_test.iloc[:,2:3]
# y_Species
Family_train = y_train.iloc[:,0:1]
# Family_train
Genus_train = y_train.iloc[:,1:2]
# X_Genus
Species_train = y_train.iloc[:,2:3]
# X_Species

### (b) Each instance has three labels: Families, Genus, and Species. Each of the labels has multiple classes. We wish to solve a multi-class and multi-label problem. One of the most important approaches to multi-label classification is to train a classifier for each label (binary relevance). We first try this approach:

### i. Research exact match and hamming score/ loss methods for evaluating multilabel classification and use them in evaluating the classifiers in this problem

### ii. Train a SVM for each of the labels, using Gaussian kernels and one versus all classifiers. Determine the weight of the SVM penalty and the width of the Gaussian Kernel using 10 fold cross validation.1 You are welcome to try to solve the problem with both standardized 2 and raw attributes and report the results.

##### one versus all classifiers

In [72]:
y_test

Family          Genus                 Species
406   Leptodactylidae      Adenomera          AdenomeraAndre
698     Dendrobatidae       Ameerega      Ameeregatrivittata
1893  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
761     Dendrobatidae       Ameerega      Ameeregatrivittata
5931          Hylidae      Hypsiboas       HypsiboasCordobae
...               ...            ...                     ...
1745  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
4922          Hylidae  Dendropsophus              HylaMinuta
2093  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
2780  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
944     Dendrobatidae       Ameerega      Ameeregatrivittata

[2158 rows x 3 columns]

In [74]:
# Family
parameters = { 'estimator__kernel':['rbf'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel='rbf'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_train, Family_train)

Family_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The Hamming loss is : %f"%hamming_loss(Family_test,Family_pred))
print("The Ecact match ration is : %f"%accuracy_score(Family_test,Family_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.5, 'estimator__kernel': 'rbf'}
The Hamming loss is : 0.000000
The Ecact match ration is : 1.000000


In [75]:
# Genus
parameters = { 'estimator__kernel':['rbf'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel = 'rbf'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_train, Genus_train)

Genus_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Genus_train))
print("The test score of rbf is : %f"%clf.score(X_test,Genus_test))
print("The Hamming loss is : %f"%hamming_loss(Genus_test,Genus_pred))
print("The Ecact match ration is : %f"%accuracy_score(Genus_test,Genus_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.5, 'estimator__kernel': 'rbf'}
The training score of rbf is : 1.000000
The test score of rbf is : 1.000000
The Hamming loss is : 0.000000
The Ecact match ration is : 1.000000


In [76]:
# Species
parameters = { 'estimator__kernel':['rbf'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel = 'rbf'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_train, Species_train)

Species_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Species_train))
print("The test score of rbf is : %f"%clf.score(X_test,Species_test))
print("The Hamming loss is : %f"%hamming_loss(Species_test,Species_pred))
print("The Ecact match ration is : %f"%accuracy_score(Species_test,Species_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 0.1, 'estimator__gamma': 1, 'estimator__kernel': 'rbf'}
The training score of rbf is : 0.999801
The test score of rbf is : 0.999537
The Hamming loss is : 0.000463
The Ecact match ration is : 0.999537


In [78]:
Family_pred =  pd.DataFrame(Family_pred)
Genus_pred =  pd.DataFrame(Genus_pred)
Species_pred =  pd.DataFrame(Species_pred)
# Species_pred

Data_Pred = pd.concat([Family_pred,Genus_pred,Species_pred],axis=1,ignore_index=True)
Data_Pred.columns = ['Family','Genus','Species']
Data_Pred

Family          Genus                 Species
0     Leptodactylidae      Adenomera          AdenomeraAndre
1       Dendrobatidae       Ameerega      Ameeregatrivittata
2     Leptodactylidae      Adenomera  AdenomeraHylaedactylus
3       Dendrobatidae       Ameerega      Ameeregatrivittata
4             Hylidae      Hypsiboas       HypsiboasCordobae
...               ...            ...                     ...
2153  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
2154          Hylidae  Dendropsophus              HylaMinuta
2155  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
2156  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
2157    Dendrobatidae       Ameerega      Ameeregatrivittata

[2158 rows x 3 columns]

In [79]:
y_test

Family          Genus                 Species
406   Leptodactylidae      Adenomera          AdenomeraAndre
698     Dendrobatidae       Ameerega      Ameeregatrivittata
1893  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
761     Dendrobatidae       Ameerega      Ameeregatrivittata
5931          Hylidae      Hypsiboas       HypsiboasCordobae
...               ...            ...                     ...
1745  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
4922          Hylidae  Dendropsophus              HylaMinuta
2093  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
2780  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
944     Dendrobatidae       Ameerega      Ameeregatrivittata

[2158 rows x 3 columns]

In [82]:
y_test['Combine']= y_test['Family']+y_test['Genus']+y_test['Species']
# y_test
Data_Pred['Combine']= Data_Pred['Family']+Data_Pred['Genus']+Data_Pred['Species']
# Data_Pred

Combine_test = y_test.iloc[:,3:]
Combine_Pred = Data_Pred.iloc[:,3:]

In [86]:
# Combine_test
Combine_Pred

Combine
0             LeptodactylidaeAdenomeraAdenomeraAndre
1            DendrobatidaeAmeeregaAmeeregatrivittata
2     LeptodactylidaeAdenomeraAdenomeraHylaedactylus
3            DendrobatidaeAmeeregaAmeeregatrivittata
4                  HylidaeHypsiboasHypsiboasCordobae
...                                              ...
2153  LeptodactylidaeAdenomeraAdenomeraHylaedactylus
2154                  HylidaeDendropsophusHylaMinuta
2155  LeptodactylidaeAdenomeraAdenomeraHylaedactylus
2156  LeptodactylidaeAdenomeraAdenomeraHylaedactylus
2157         DendrobatidaeAmeeregaAmeeregatrivittata

[2158 rows x 1 columns]

In [81]:
print("The Hamming loss is : %f"%hamming_loss(Combine_test,Combine_Pred))
print("The Ecact match ration is : %f"%accuracy_score(Combine_test,Combine_Pred))

The Hamming loss is : 0.000463
The Ecact match ration is : 0.999537


### iii. Repeat 1(b)ii with L1-penalized SVMs.3 Remember to standardize the attributes. Determine the weight of the SVM penalty using 10 fold cross validation.

In [93]:
# standardize the attributes
std_X_train = preprocessing.scale(X_train)
std_X_train = pd.DataFrame(std_X_train)
std_X_train

0         1         2         3         4         5         6   \
0     0.140814 -0.810665 -0.658931  0.632074  0.626255 -0.046444 -0.690079   
1     0.140814 -0.181504  1.891570 -1.914634 -1.481353  1.088392  0.554112   
2     0.140814 -1.146937 -0.947419  0.750919  0.697752  0.124207 -0.589863   
3     0.140814  0.303043 -0.275622  0.938919  0.397757 -0.870912 -0.855762   
4     0.140814  0.244305  0.499332  1.067859  0.041488 -0.462582 -0.047774   
...        ...       ...       ...       ...       ...       ...       ...   
5032  0.140814 -0.035903  0.027968  0.581602  0.175391  0.209846 -0.134616   
5033  0.140814 -1.787392 -1.541120  0.432815  1.133616  0.352788  0.254294   
5034  0.140814  0.767315  0.209739  0.592844  0.002793  0.330008  0.263065   
5035  0.140814 -0.587039 -1.357364 -1.995117  0.366253  0.748851  0.579131   
5036  0.140814 -0.636455  1.355460 -0.248526 -2.382757  1.621031  1.896695   

            7         8         9   ...        13        14        15  \
0    -0.420477  0.176641  0.032989  ... -1.802038 -0.067072  1.824920   
1     1.055589 -1.455240 -1.550939  ...  1.374288  0.252699 -2.584654   
2    -0.170062  0.631853  0.490652  ...  0.459862 -1.077281 -0.914801   
3     0.336108  0.907654  0.110578  ...  0.284166 -0.676544  0.014032   
4     0.576564  1.022278 -0.788784  ... -0.365493 -0.955396  0.369371   
...        ...       ...       ...  ...       ...       ...       ...   
5032  0.521997  0.611792 -1.083995  ... -0.654973 -0.883693  0.877387   
5033 -1.197094 -0.646751  2.617946  ...  2.668256  1.835492 -1.696955   
5034  0.777945  0.373253 -1.463072  ... -1.193266 -0.899354  0.679962   
5035 -0.440305 -0.765207  1.262913  ...  0.972588  1.934187  0.775846   
5036 -1.688601 -0.750950  0.763211  ...  1.097424 -0.552204 -0.980064   

            16        17        18        19        20        21        22  
0     0.105110 -1.717788 -0.791900  0.404000  1.405875  0.703569 -0.310811  
1    -0.700735  1.088214  1.177197  1.816278  0.220708 -0.772414  2.583009  
2     1.189631  2.077716  0.390768 -1.597754 -1.910231  0.966604 -0.310811  
3     1.274097  1.529478  0.552454 -0.529486  0.008642  0.852478 -0.386964  
4     0.196313 -0.576706 -0.393680 -0.614720 -0.433265  0.531213 -0.463118  
...        ...       ...       ...       ...       ...       ...       ...  
5032  0.793779 -0.808748 -1.104080 -0.319712  1.013592  0.558078 -0.463118  
5033 -2.214445 -0.634019  1.758953  2.895063  0.897872 -1.806556  0.679180  
5034  0.461407 -1.304713 -0.835215  0.448807  1.649967  0.763941 -0.463118  
5035 -0.795614 -1.591828 -1.064928  0.354568  0.708994 -0.383511 -1.453109  
5036  0.463351  0.837880 -0.215836  0.607078  1.020399 -0.833447  1.669171  

[5037 rows x 23 columns]

In [100]:
# Family
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel='linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_train, Family_train)

Family_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The Hamming loss is : %f"%hamming_loss(Family_test,Family_pred))
print("The Ecact match ration is : %f"%accuracy_score(Family_test,Family_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.005, 'estimator__kernel': 'linear'}
The Hamming loss is : 0.039852
The Ecact match ration is : 0.960148


In [99]:
# Family_pred_L1

In [101]:
# Genus
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel = 'linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_train, Genus_train)

Genus_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Genus_train))
print("The test score of rbf is : %f"%clf.score(X_test,Genus_test))
print("The Hamming loss is : %f"%hamming_loss(Genus_test,Genus_pred))
print("The Ecact match ration is : %f"%accuracy_score(Genus_test,Genus_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.005, 'estimator__kernel': 'linear'}
The training score of rbf is : 0.965654
The test score of rbf is : 0.955051
The Hamming loss is : 0.044949
The Ecact match ration is : 0.955051


In [103]:
# Species
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel = 'linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_train, Species_train)

Species_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Species_train))
print("The test score of rbf is : %f"%clf.score(X_test,Species_test))
print("The Hamming loss is : %f"%hamming_loss(Species_test,Species_pred))
print("The Ecact match ration is : %f"%accuracy_score(Species_test,Species_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.005, 'estimator__kernel': 'linear'}
The training score of rbf is : 0.972603
The test score of rbf is : 0.965246
The Hamming loss is : 0.034754
The Ecact match ration is : 0.965246


In [105]:
Family_pred =  pd.DataFrame(Family_pred)
Genus_pred =  pd.DataFrame(Genus_pred)
Species_pred =  pd.DataFrame(Species_pred)
# Species_pred

Data_Pred = pd.concat([Family_pred,Genus_pred,Species_pred],axis=1,ignore_index=True)
Data_Pred.columns = ['Family','Genus','Species']
Data_Pred

y_test['Combine']= y_test['Family']+y_test['Genus']+y_test['Species']
# y_test
Data_Pred['Combine']= Data_Pred['Family']+Data_Pred['Genus']+Data_Pred['Species']
# Data_Pred

Combine_test = y_test.iloc[:,3:]
Combine_Pred = Data_Pred.iloc[:,3:]

In [106]:
print("The Hamming loss is : %f"%hamming_loss(Combine_test,Combine_Pred))
print("The Ecact match ration is : %f"%accuracy_score(Combine_test,Combine_Pred))

The Hamming loss is : 0.065338
The Ecact match ration is : 0.934662


### iv. Repeat 1(b)iii by using SMOTE or any other method you know to remedy class imbalance. Report your conclusions about the classifiers you trained.

In [14]:
sm = SMOTE(random_state=11)
X_sm_famliy, Family_sm = sm.fit_resample(X_train, Family_train)

In [15]:
sm = SMOTE(random_state=11)
X_sm_genus, Genus_sm = sm.fit_resample(X_train, Genus_train)

In [16]:
sm = SMOTE(random_state=11)
X_sm_species, Species_sm = sm.fit_resample(X_train, Species_train)

In [15]:
# Family
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel='linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_sm_famliy, Family_sm)

Family_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The Hamming loss is : %f"%hamming_loss(Family_test,Family_pred))
print("The Ecact match ration is : %f"%accuracy_score(Family_test,Family_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.005, 'estimator__kernel': 'linear'}
The Hamming loss is : 0.108434
The Ecact match ration is : 0.891566


In [11]:
# Genus
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel = 'linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_sm_genus, Genus_sm)

Genus_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Genus_train))
print("The test score of rbf is : %f"%clf.score(X_test,Genus_test))
print("The Hamming loss is : %f"%hamming_loss(Genus_test,Genus_pred))
print("The Ecact match ration is : %f"%accuracy_score(Genus_test,Genus_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits
{'estimator__C': 1, 'estimator__gamma': 0.005, 'estimator__kernel': 'linear'}
The training score of rbf is : 0.910066
The test score of rbf is : 0.906858
The Hamming loss is : 0.093142
The Ecact match ration is : 0.906858


In [17]:
# Species
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.1,1],'estimator__gamma':[0.005,0.05]}

OVR = OneVsRestClassifier(SVC(kernel = 'linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_sm_species, Species_sm)

Species_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Species_train))
print("The test score of rbf is : %f"%clf.score(X_test,Species_test))
print("The Hamming loss is : %f"%hamming_loss(Species_test,Species_pred))
print("The Ecact match ration is : %f"%accuracy_score(Species_test,Species_pred))

Fitting 10 folds for each of 4 candidates, totalling 40 fits
{'estimator__C': 1, 'estimator__gamma': 0.005, 'estimator__kernel': 'linear'}
The training score of rbf is : 0.953941
The test score of rbf is : 0.951344
The Hamming loss is : 0.048656
The Ecact match ration is : 0.951344


In [ ]:
# Species
parameters = { 'estimator__kernel':['linear'],'estimator__C': [0.0001,0.001,0.01,0.1,1],'estimator__gamma':[0.005,0.05,0.5,1,10,100]}

OVR = OneVsRestClassifier(SVC(kernel = 'linear'),n_jobs = -1)
clf =  GridSearchCV(OVR,parameters,cv=10,verbose=3,n_jobs = -1)

clf.fit(X_sm_species, Species_sm)

Species_pred = clf.predict(X_test)

best_para = clf.best_params_
print(best_para)
# print("The best parameter is : %f"%{best_para})
print("The training score of rbf is : %f"%clf.score(X_train,Species_train))
print("The test score of rbf is : %f"%clf.score(X_test,Species_test))
print("The Hamming loss is : %f"%hamming_loss(Species_test,Species_pred))
print("The Ecact match ration is : %f"%accuracy_score(Species_test,Species_pred))

Fitting 10 folds for each of 30 candidates, totalling 300 fits


### 2. K-Means Clustering on a Multi-Class and Multi-Label Data Set Monte-Carlo Simulation: Perform the following procedures 50 times, and report the average and standard deviation of the 50 Hamming Distances that you calculate.

### (a) Use k-means clustering on the whole Anuran Calls (MFCCs) Data Set (do not split the data into train and test, as we are not performing supervised learning in this exercise). Choose k ∈ {1, 2, . . . , 50} automatically based on one of the methods provided in the slides (CH or Gap Statistics or scree plots or Silhouettes) or any other method you know.

In [24]:
# y_train
# X_train
# y_test
# X_test

RawData

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
...        ...       ...       ...       ...       ...       ...       ...   
7190       1.0 -0.554504 -0.337717  0.035533  0.034511  0.443451  0.093889   
7191       1.0 -0.517273 -0.370574  0.030673  0.068097  0.402890  0.096628   
7192       1.0 -0.582557 -0.343237  0.029468  0.064179  0.385596  0.114905   
7193       1.0 -0.519497 -0.307553 -0.004922  0.072865  0.377131  0.086866   
7194       1.0 -0.508833 -0.324106  0.062068  0.078211  0.397188  0.094596   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_17  MFCCs_18  MFCCs_19  \
0    -0.150063 -0.171128  0.124676  ... -0.108351 -0.077623 -0.009568   
1    -0.222475 -0.207693  0.170883  ... -0.090974 -0.056510 -0.035303   
2    -0.242234 -0.219153  0.232538  ... -0.050691 -0.023590 -0.066722   
3    -0.194347 -0.098181  0.270375  ... -0.136009 -0.177037 -0.130498   
4    -0.265423 -0.172700  0.266434  ... -0.048885 -0.053074 -0.088550   
...        ...       ...       ...  ...       ...       ...       ...   
7190 -0.100753  0.037087  0.081075  ...  0.069430  0.071001  0.021591   
7191 -0.116460  0.063727  0.089034  ...  0.061127  0.068978  0.017745   
7192 -0.103317  0.070370  0.081317  ...  0.082474  0.077771 -0.009688   
7193 -0.115799  0.056979  0.089316  ...  0.051796  0.069073  0.017963   
7194 -0.117672  0.058874  0.076180  ...  0.061455  0.072983 -0.003980   

      MFCCs_20  MFCCs_21  MFCCs_22           Family      Genus  \
0     0.057684  0.118680  0.014038  Leptodactylidae  Adenomera   
1     0.020140  0.082263  0.029056  Leptodactylidae  Adenomera   
2    -0.025083  0.099108  0.077162  Leptodactylidae  Adenomera   
3    -0.054766 -0.018691  0.023954  Leptodactylidae  Adenomera   
4    -0.031346  0.108610  0.079244  Leptodactylidae  Adenomera   
...        ...       ...       ...              ...        ...   
7190  0.052449 -0.021860 -0.079860          Hylidae     Scinax   
7191  0.046461 -0.015418 -0.101892          Hylidae     Scinax   
7192  0.027834 -0.000531 -0.080425          Hylidae     Scinax   
7193  0.041803 -0.027911 -0.096895          Hylidae     Scinax   
7194  0.031560 -0.029355 -0.087910          Hylidae     Scinax   

             Species  RecordID  
0     AdenomeraAndre         1  
1     AdenomeraAndre         1  
2     AdenomeraAndre         1  
3     AdenomeraAndre         1  
4     AdenomeraAndre         1  
...              ...       ...  
7190     ScinaxRuber        60  
7191     ScinaxRuber        60  
7192     ScinaxRuber        60  
7193     ScinaxRuber        60  
7194     ScinaxRuber        60  

[7195 rows x 26 columns]

In [85]:
# Data_Pred = pd.concat([Family_pred,Genus_pred,Species_pred],axis=1,ignore_index=True)
X = pd.concat([X_train,X_test],ignore_index=True)
y = pd.concat([y_train,y_test],ignore_index=True)

# X
# y

In [87]:
for k in range (2,51):
    KMean= KMeans(n_clusters=k,random_state = 42)
    KMean.fit(X)
    label=KMean.predict(X)
    print(f'k=',k)
    print(f'Silhouette Score(n=k): {silhouette_score(X, label)}')

k= 2
Silhouette Score(n=k): 0.34867784102771515
k= 3
Silhouette Score(n=k): 0.36768245219926315
k= 4
Silhouette Score(n=k): 0.37867825251598397
k= 5
Silhouette Score(n=k): 0.37149147864771287
k= 6
Silhouette Score(n=k): 0.26400802111671495
k= 7
Silhouette Score(n=k): 0.26037818747186503
k= 8
Silhouette Score(n=k): 0.2806155554940955
k= 9
Silhouette Score(n=k): 0.27643555164562594
k= 10
Silhouette Score(n=k): 0.2622417506439173
k= 11
Silhouette Score(n=k): 0.2690019279474747
k= 12
Silhouette Score(n=k): 0.2719536334592367
k= 13
Silhouette Score(n=k): 0.278298660340801
k= 14
Silhouette Score(n=k): 0.2657977325098931
k= 15
Silhouette Score(n=k): 0.26971669145635296
k= 16
Silhouette Score(n=k): 0.28650229843618247
k= 17
Silhouette Score(n=k): 0.25743354487654313
k= 18
Silhouette Score(n=k): 0.2787816593270135
k= 19
Silhouette Score(n=k): 0.27865066011889883
k= 20
Silhouette Score(n=k): 0.2663809259718965
k= 21
Silhouette Score(n=k): 0.2748180536949199
k= 22
Silhouette Score(n=k): 0.2613421

The best k is 4. And the Silhouette Score(n=k): 0.3787509343305295

### (b) In each cluster, determine which family is the majority by reading the true labels. Repeat for genus and species.

In [114]:
# y
Family = y.iloc[:,0:1]
Genus = y.iloc[:,1:2]
Species = y.iloc[:,2:3]
# Family
# Genus
# Species

In [162]:
kmeans= KMeans(n_clusters=4,random_state = 42)
kmeans.fit(X)
label_k4=kmeans.predict(X)

print(f'Silhouette Score(n=4): {silhouette_score(X, label)}')
label_k4 = pd.DataFrame(label_k4)
label_k4.columns = ['kmeans_k4']
# KMean.labels_
label_k4

D:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Silhouette Score(n=4): 0.21901843643523833


kmeans_k4
0             0
1             3
2             3
3             3
4             0
...         ...
7190          3
7191          0
7192          1
7193          3
7194          0

[7195 rows x 1 columns]

In [118]:
label_k4.groupby(['kmeans_k4'],as_index=False)['kmeans_k4'].agg({'cnt':'count'})

kmeans_k4   cnt
0          0  3568
1          1  1032
2          2   615
3          3  1980

In [115]:
Family.groupby(['Family'],as_index=False)['Family'].agg({'cnt':'count'})

Family   cnt
0        Bufonidae    68
1    Dendrobatidae   542
2          Hylidae  2165
3  Leptodactylidae  4420

In [116]:
Genus.groupby(['Genus'],as_index=False)['Genus'].agg({'cnt':'count'})

Genus   cnt
0      Adenomera  4150
1       Ameerega   542
2  Dendropsophus   310
3      Hypsiboas  1593
4  Leptodactylus   270
5  Osteocephalus   114
6       Rhinella    68
7         Scinax   148

In [117]:
Species.groupby(['Species'],as_index=False)['Species'].agg({'cnt':'count'})

Species   cnt
0          AdenomeraAndre   672
1  AdenomeraHylaedactylus  3478
2      Ameeregatrivittata   542
3              HylaMinuta   310
4    HypsiboasCinerascens   472
5       HypsiboasCordobae  1121
6     LeptodactylusFuscus   270
7   OsteocephalusOophagus   114
8       Rhinellagranulosa    68
9             ScinaxRuber   148

In [128]:
# groupby([ ]).size()
Family_kmeans = pd.concat([label_k4, Family], axis=1)
# Family_kmeans
# Family_kmeans.groupby(['Species']).size().sort_values(ascending=False)
Family_kmeans.groupby(['kmeans_k4','Family']).size().sort_values(ascending=False)

kmeans_k4  Family         
0          Leptodactylidae    3467
3          Hylidae            1245
           Leptodactylidae     628
2          Hylidae             590
1          Dendrobatidae       501
           Leptodactylidae     302
           Hylidae             229
0          Hylidae             101
3          Bufonidae            66
           Dendrobatidae        41
2          Leptodactylidae      23
           Bufonidae             2
dtype: int64

In [129]:
# groupby([ ]).size()
Genus_kmeans = pd.concat([label_k4, Genus], axis=1)
Genus_kmeans.groupby(['kmeans_k4','Genus']).size().sort_values(ascending=False)

kmeans_k4  Genus        
0          Adenomera        3466
3          Hypsiboas        1038
2          Hypsiboas         542
1          Ameerega          501
3          Adenomera         375
1          Adenomera         296
3          Leptodactylus     253
1          Dendropsophus     189
3          Scinax             96
0          Dendropsophus      86
3          Osteocephalus      76
           Rhinella           66
           Ameerega           41
1          Scinax             39
2          Osteocephalus      38
3          Dendropsophus      35
2          Adenomera          13
0          Hypsiboas          12
2          Scinax             10
           Leptodactylus      10
1          Leptodactylus       6
0          Scinax              3
2          Rhinella            2
1          Hypsiboas           1
0          Leptodactylus       1
dtype: int64

In [130]:
# groupby([ ]).size()
Species_kmeans = pd.concat([label_k4, Species], axis=1)
Species_kmeans.groupby(['kmeans_k4','Species']).size().sort_values(ascending=False)

kmeans_k4  Species               
0          AdenomeraHylaedactylus    3466
3          HypsiboasCordobae         1018
1          Ameeregatrivittata         501
2          HypsiboasCinerascens       452
3          AdenomeraAndre             370
1          AdenomeraAndre             296
3          LeptodactylusFuscus        253
1          HylaMinuta                 189
3          ScinaxRuber                 96
2          HypsiboasCordobae           90
0          HylaMinuta                  86
3          OsteocephalusOophagus       76
           Rhinellagranulosa           66
           Ameeregatrivittata          41
1          ScinaxRuber                 39
2          OsteocephalusOophagus       38
3          HylaMinuta                  35
           HypsiboasCinerascens        20
0          HypsiboasCordobae           12
2          LeptodactylusFuscus         10
           ScinaxRuber                 10
           AdenomeraHylaedactylus       7
1          LeptodactylusFuscus          6


For Label 0:  
    Family: Leptodactylidae(3467)  
    Genus: Adenomera(3466)    
    Species: AdenomeraHylaedactylus(3466)  
       
For Label 1:  
    Family: Dendrobatidae(501)  
    Genus: Ameerega(501)  
    Species: Ameeregatrivittata(501)  
      
For Label 2:  
    Family: Hylidae(590)  
    Genus: Hypsiboas(542)  
    Species: HypsiboasCinerascens(452)  
      
For Label 3:  
    Family: Hylidae(1245)  
    Genus: Hypsiboas(1038)  
    Species: HypsiboasCordobae(1018)  

### (c) Now for each cluster you have a majority label triplet (family, genus, species). Calculate the average Hamming distance, Hamming score, and Hamming loss between the true labels and the labels assigned by clusters.

In [166]:
# # label_k4
# label_k4_family = label_k4.rename(columns={'kmeans_k4': 'label_k4_family'})
# label_k4_family['label_k4_family'] = label_k4_family.Category.map(lambda x: 'Leptodactylidae' if x ==0)

In [179]:
label_k4_family = label_k4.copy(deep=True)
# print(label_k4_family.columns)
# label_k4_family.rename(columns={'kmeans_k4':'label_k4_family'})
for i in range(len(label_k4_family['kmeans_k4'])):   
#     print(label_k4['kmeans_k4'][i])
    if label_k4_family['kmeans_k4'][i] == 0:
        label_k4_family['kmeans_k4'][i] = 'Leptodactylidae'
    
    if label_k4_family['kmeans_k4'][i] == 1:
        label_k4_family['kmeans_k4'][i] = 'Dendrobatidae'

    if label_k4_family['kmeans_k4'][i] == 2:
        label_k4_family['kmeans_k4'][i] = 'Hylidae'
#     print(label_k4['kmeans_k4'][i])
    if label_k4_family['kmeans_k4'][i] == 3:
        label_k4_family['kmeans_k4'][i] = 'Hylidae'
        
#     print(label_k4['kmeans_k4'][i])
label_k4_family.rename(columns={'kmeans_k4':'label_k4_family'},inplace = True)
label_k4_family

label_k4_family
0     Leptodactylidae
1             Hylidae
2             Hylidae
3             Hylidae
4     Leptodactylidae
...               ...
7190          Hylidae
7191  Leptodactylidae
7192    Dendrobatidae
7193          Hylidae
7194  Leptodactylidae

[7195 rows x 1 columns]

In [181]:
label_k4_Genus = label_k4.copy(deep=True)
# label_k4_Genus = label_k4.rename(columns={'kmeans_k4': 'label_k4_Genus'})
for i in range(len(label_k4_Genus['kmeans_k4'])):   
#     print(label_k4['kmeans_k4'][i])
    if label_k4_Genus['kmeans_k4'][i] == 0:
        label_k4_Genus['kmeans_k4'][i] = 'Adenomera'
    
    if label_k4_Genus['kmeans_k4'][i] == 1:
        label_k4_Genus['kmeans_k4'][i] = 'Ameerega'

    if label_k4_Genus['kmeans_k4'][i] == 2:
        label_k4_Genus['kmeans_k4'][i] = 'Hypsiboas'
#     print(label_k4['kmeans_k4'][i])
    if label_k4_Genus['kmeans_k4'][i] == 3:
        label_k4_Genus['kmeans_k4'][i] = 'Hypsiboas'
        
#     print(label_k4['kmeans_k4'][i])
label_k4_Genus.rename(columns={'kmeans_k4':'label_k4_Genus'},inplace = True)
label_k4_Genus

label_k4_Genus
0         Adenomera
1         Hypsiboas
2         Hypsiboas
3         Hypsiboas
4         Adenomera
...             ...
7190      Hypsiboas
7191      Adenomera
7192       Ameerega
7193      Hypsiboas
7194      Adenomera

[7195 rows x 1 columns]

In [182]:
label_k4_Species = label_k4.copy(deep=True)
# label_k4_Genus = label_k4.rename(columns={'kmeans_k4': 'label_k4_Genus'})
for i in range(len(label_k4_Species['kmeans_k4'])):   
#     print(label_k4['kmeans_k4'][i])
    if label_k4_Species['kmeans_k4'][i] == 0:
        label_k4_Species['kmeans_k4'][i] = 'AdenomeraHylaedactylus'
    
    if label_k4_Species['kmeans_k4'][i] == 1:
        label_k4_Species['kmeans_k4'][i] = 'Ameeregatrivittata'

    if label_k4_Species['kmeans_k4'][i] == 2:
        label_k4_Species['kmeans_k4'][i] = 'HypsiboasCinerascens'
#     print(label_k4['kmeans_k4'][i])
    if label_k4_Species['kmeans_k4'][i] == 3:
        label_k4_Species['kmeans_k4'][i] = 'HypsiboasCordobae'
        
#     print(label_k4['kmeans_k4'][i])
label_k4_Species.rename(columns={'kmeans_k4':'label_k4_Species'},inplace = True)
label_k4_Species

label_k4_Species
0     AdenomeraHylaedactylus
1          HypsiboasCordobae
2          HypsiboasCordobae
3          HypsiboasCordobae
4     AdenomeraHylaedactylus
...                      ...
7190       HypsiboasCordobae
7191  AdenomeraHylaedactylus
7192      Ameeregatrivittata
7193       HypsiboasCordobae
7194  AdenomeraHylaedactylus

[7195 rows x 1 columns]

In [185]:
label_k4_fgs = pd.concat([label_k4_family,label_k4_Genus,label_k4_Species], axis=1)
label_k4_fgs

label_k4_family label_k4_Genus        label_k4_Species
0     Leptodactylidae      Adenomera  AdenomeraHylaedactylus
1             Hylidae      Hypsiboas       HypsiboasCordobae
2             Hylidae      Hypsiboas       HypsiboasCordobae
3             Hylidae      Hypsiboas       HypsiboasCordobae
4     Leptodactylidae      Adenomera  AdenomeraHylaedactylus
...               ...            ...                     ...
7190          Hylidae      Hypsiboas       HypsiboasCordobae
7191  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
7192    Dendrobatidae       Ameerega      Ameeregatrivittata
7193          Hylidae      Hypsiboas       HypsiboasCordobae
7194  Leptodactylidae      Adenomera  AdenomeraHylaedactylus

[7195 rows x 3 columns]

In [364]:
label_k4_origin = pd.concat([label_k4_fgs,y], axis=1)
label_k4_origin.sort_values("label_k4_Species",inplace=True)
label_k4_origin

label_k4_family label_k4_Genus        label_k4_Species           Family  \
0     Leptodactylidae      Adenomera  AdenomeraHylaedactylus  Leptodactylidae   
3990  Leptodactylidae      Adenomera  AdenomeraHylaedactylus  Leptodactylidae   
3989  Leptodactylidae      Adenomera  AdenomeraHylaedactylus  Leptodactylidae   
3986  Leptodactylidae      Adenomera  AdenomeraHylaedactylus  Leptodactylidae   
3985  Leptodactylidae      Adenomera  AdenomeraHylaedactylus  Leptodactylidae   
...               ...            ...                     ...              ...   
1061          Hylidae      Hypsiboas       HypsiboasCordobae          Hylidae   
5018          Hylidae      Hypsiboas       HypsiboasCordobae          Hylidae   
5019          Hylidae      Hypsiboas       HypsiboasCordobae          Hylidae   
5023          Hylidae      Hypsiboas       HypsiboasCordobae          Hylidae   
2375          Hylidae      Hypsiboas       HypsiboasCordobae          Hylidae   

              Genus                 Species  
0         Adenomera  AdenomeraHylaedactylus  
3990      Adenomera  AdenomeraHylaedactylus  
3989      Adenomera  AdenomeraHylaedactylus  
3986      Adenomera  AdenomeraHylaedactylus  
3985      Adenomera  AdenomeraHylaedactylus  
...             ...                     ...  
1061  Osteocephalus   OsteocephalusOophagus  
5018      Hypsiboas       HypsiboasCordobae  
5019      Hypsiboas       HypsiboasCordobae  
5023      Hypsiboas       HypsiboasCordobae  
2375      Hypsiboas       HypsiboasCordobae  

[7195 rows x 6 columns]

In [358]:
# label_k4_origin.loc[:,'label_k4_family'] = '0'
# label_k4_origin.loc[:,'label_k4_Genus'] = '0'
# label_k4_origin.loc[:,'label_k4_Species'] = '0'
# # label_k4_origin

In [365]:
# Label_0_pred = label_k4_origin.iloc[:,0:3]
# Label_0_pred

In [367]:
Label_0 = label_k4_origin.loc[label_k4_origin['label_k4_Species']=='AdenomeraHylaedactylus']
Label_1 = label_k4_origin.loc[label_k4_origin['label_k4_Species']=='Ameeregatrivittata']
Label_2 = label_k4_origin.loc[label_k4_origin['label_k4_Species']=='HypsiboasCinerascens']
Label_3 = label_k4_origin.loc[label_k4_origin['label_k4_Species']=='HypsiboasCordobae']

# Label_0

In [359]:
# Label_0

In [295]:
# Label_0_int = Label_0.copy(deep=True)


# # replacement_mapping_dict = {
# #     "Leptodactylidae": 1,
# #     "Hylidae": 0,
# #     "Dendrobatidae": 0,
# #     "Bufonidae": 0
# # }
# Label_0_int["Family"].replace('Label_0_int',1)
# # Label_0_int = pd.DataFrame(Label_0_int)
# Label_0_int


In [387]:
Label_0_pred = Label_0.iloc[:,0:3]
Label_0_true = Label_0.iloc[:,3:6]
Label_1_pred = Label_1.iloc[:,0:3]
Label_1_true = Label_1.iloc[:,3:6]
Label_2_pred = Label_2.iloc[:,0:3]
Label_2_true = Label_2.iloc[:,3:6]
Label_3_pred = Label_3.iloc[:,0:3]
Label_3_true = Label_3.iloc[:,3:6]

Label_0_pred.loc[:,'label_k4_family'] = '0'
Label_0_pred.loc[:,'label_k4_Genus'] = '0'
Label_0_pred.loc[:,'label_k4_Species'] = '0'

Label_1_pred.loc[:,'label_k4_family'] = '2'
Label_1_pred.loc[:,'label_k4_Genus'] = '5'
Label_1_pred.loc[:,'label_k4_Species'] = '6'

Label_2_pred.loc[:,'label_k4_family'] = '1'
Label_2_pred.loc[:,'label_k4_Genus'] = '1'
Label_2_pred.loc[:,'label_k4_Species'] = '7'

Label_3_pred.loc[:,'label_k4_family'] = '1'
Label_3_pred.loc[:,'label_k4_Genus'] = '1'
Label_3_pred.loc[:,'label_k4_Species'] = '1'

C:\Users\luoli\AppData\Local\Temp\ipykernel_5216\830522494.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Label_0_pred.loc[:,'label_k4_family'] = '0'
C:\Users\luoli\AppData\Local\Temp\ipykernel_5216\830522494.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Label_0_pred.loc[:,'label_k4_Genus'] = '0'
C:\Users\luoli\AppData\Local\Temp\ipykernel_5216\830522494.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = v

In [372]:
pred_all = pd.concat([Label_0_pred, Label_1_pred, Label_2_pred,Label_3_pred])
pred_all

label_k4_family label_k4_Genus label_k4_Species
0                  0              0                0
3990               0              0                0
3989               0              0                0
3986               0              0                0
3985               0              0                0
...              ...            ...              ...
1061               3              3                3
5018               3              3                3
5019               3              3                3
5023               3              3                3
2375               3              3                3

[7195 rows x 3 columns]

In [373]:
true_all = pd.concat([Label_0_true, Label_1_true, Label_2_true,Label_3_true])
true_all

Family          Genus                 Species
0     Leptodactylidae      Adenomera  AdenomeraHylaedactylus
3990  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
3989  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
3986  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
3985  Leptodactylidae      Adenomera  AdenomeraHylaedactylus
...               ...            ...                     ...
1061          Hylidae  Osteocephalus   OsteocephalusOophagus
5018          Hylidae      Hypsiboas       HypsiboasCordobae
5019          Hylidae      Hypsiboas       HypsiboasCordobae
5023          Hylidae      Hypsiboas       HypsiboasCordobae
2375          Hylidae      Hypsiboas       HypsiboasCordobae

[7195 rows x 3 columns]

In [452]:
true_all.dtypes

Family     int64
Genus      int64
Species    int64
dtype: object

In [451]:
# pred_all.dtypes
pred_all["label_k4_family"]=pred_all["label_k4_family"].astype(str).astype(int)
pred_all["label_k4_Genus"]=pred_all["label_k4_Genus"].astype(str).astype(int)
pred_all["label_k4_Species"]=pred_all["label_k4_Species"].astype(str).astype(int)
pred_all.dtypes

label_k4_family     int32
label_k4_Genus      int32
label_k4_Species    int32
dtype: object

In [388]:
labels = ['Family','Genus','Species']

for i in labels:
    true_all[i],_=pd.factorize(true_all[i])
true_all

# true_all.groupby(['Family'],as_index=False)['Family'].agg({'cnt':'count'})
# true_all.groupby(['Genus'],as_index=False)['Genus'].agg({'cnt':'count'})
# true_all.groupby(['Species'],as_index=False)['Species'].agg({'cnt':'count'})

# Label_0_true.groupby(['Family'],as_index=False)['Family'].agg({'cnt':'count'})
# Label_0_true.groupby(['Genus'],as_index=False)['Genus'].agg({'cnt':'count'})
# Label_0_true.groupby(['Species'],as_index=False)['Species'].agg({'cnt':'count'})

# Label_0_true_2['Combine']= Label_0_true_2['Family']+Label_0_true_2['Genus']+Label_0_true_2['Species']
# Label_0_true_2
# # Label_0_true_combine = Label_0_true_2.iloc[:,3:4]
# # Label_0_true_combine

Family  Genus  Species
0          0      0        0
3990       0      0        0
3989       0      0        0
3986       0      0        0
3985       0      0        0
...      ...    ...      ...
1061       1      6        8
5018       1      1        1
5019       1      1        1
5023       1      1        1
2375       1      1        1

[7195 rows x 3 columns]

In [453]:
from sklearn.metrics import hamming_loss
import numpy as np

hamming_distance_results = []
hamming_loss_results = []
hamming_score_results = [] 

for index, row in pred_all.iterrows():
    df1= pd.DataFrame([row["label_k4_family"],row["label_k4_Genus"],row["label_k4_Species"]])
    df1 = pd.DataFrame(df1)
    
# df1
# print(type(df1))

    for index, row in true_all.iterrows():
        df2= pd.DataFrame([row["Family"],row["Genus"],row["Species"]])
        df2 = pd.DataFrame(df2)

        hamming(df1, df2) * len(df1)
        hamming_distance_df.append(hamming)
        
#         df1_array = np.array(df1)
#         df2_array = np.array(df2)

#         df1_series = pd.Series
#         df2_series = pd.Series
    
        hamming_loss_result = hamming_loss(df1.values, df2.values)
        hamming_loss_df.append(hamming_loss_result)
        
        hamming_score_result = 1- hamming_loss_result
        hamming_score_df.append(hamming_score_result)

KeyboardInterrupt: 

### References:
   https://medium.datadriveninvestor.com/a-survey-of-evaluation-metrics-for-multilabel-classification-bb16e8cd41cd
   
   https://github.com/christianversloot/machine-learning-articles/blob/main/how-to-create-a-multilabel-svm-classifier-with-scikit-learn.md
   
   https://stackoverflow.com/questions/32239577/getting-the-accuracy-for-multi-label-prediction-in-scikit-learn
   
   https://mmuratarat.github.io/2020-01-25/multilabel_classification_metrics
   
   https://stats.stackexchange.com/questions/233275/multilabel-classification-metrics-on-scikit
   
   https://towardsdatascience.com/multiclass-classification-with-support-vector-machines-svm-kernel-trick-kernel-functions-f9d5377d6f02
   
   https://scikit-learn.org/stable/modules/generated/sklearn.multiclass.OneVsRestClassifier.html
   
   https://medium.com/all-things-ai/in-depth-parameter-tuning-for-svc-758215394769
   
   https://scikit-learn.org/stable/modules/multiclass.html
   
   https://www.statology.org/hamming-distance-python/
   
   https://www.statology.org/hamming-distance-python/
   
   https://stackoverflow.com/questions/70326263/calculating-hamming-distance-in-a-given-year